alt-tab-macos/ui/PreferencesPanel.swift

```diff
@@ -1,110 +1,211 @@
+        let input = NSTextField(string: Preferences.rawValues[rawName]!)
+        input.widthAnchor.constraint(equalToConstant: 32).isActive = true
+
         self[keyPath: keyPath] = input
-        return [label, input]
+
+        return makeLabelWithProvidedControl(labelText, rawName, input, suffixText)
     }
 
-    @objc func dropdownDidChange(sender: AnyObject) throws {
-        if let popUpButton = sender as? NSPopUpButton {
-            switch popUpButton {
-            case theme:
-                try! Preferences.updateAndValidateFromString("theme", popUpButton.titleOfSelectedItem!)
-            case metaKey:
-                try! Preferences.updateAndValidateFromString("metaKey", popUpButton.titleOfSelectedItem!)
-            case showOnScreen:
-                try! Preferences.updateAndValidateFromString("showOnScreen", popUpButton.titleOfSelectedItem!)
-            default:
-                throw "Tried to update an unknown popUpButton: '\(popUpButton)' = '\(popUpButton.titleOfSelectedItem!)'"
-            }
-            try! Preferences.saveRawToDisk()
+    private func makeLabelWithDropdown(_ keyPath: ReferenceWritableKeyPath<PreferencesPanel, NSPopUpButton?>, _ labelText: String, rawName: String, values: [String], suffixText: String? = nil) -> NSStackView {
+        let popUp = NSPopUpButton()
+        popUp.addItems(withTitles: values)
+        popUp.selectItem(withTitle: Preferences.rawValues[rawName]!)
+
+        self[keyPath: keyPath] = popUp
+
+        return makeLabelWithProvidedControl(labelText, rawName, popUp, suffixText)
+    }
+
+    private func makeLabelWithSlider(_ keyPath: ReferenceWritableKeyPath<PreferencesPanel, NSSlider?>, _ labelText: String, rawName: String, minValue: Double, maxValue: Double, numberOfTickMarks: Int, suffixText: String? = nil) -> NSStackView {
+        let slider = NSSlider(
+                value: Double(Preferences.rawValues[rawName]!)!,
+                minValue: minValue,
+                maxValue: maxValue,
+                target: self,
+                action: #selector(controlDidEndEditing)
+        )
+        slider.numberOfTickMarks = numberOfTickMarks
+        slider.allowsTickMarkValuesOnly = numberOfTickMarks > 1
+        slider.tickMarkPosition = .below
+
+        self[keyPath: keyPath] = slider
+
+        return makeLabelWithProvidedControl(labelText, rawName, slider, suffixText)
+    }
+
+    private func makeLabelWithProvidedControl(_ labelText: String, _ rawName: String, _ control: NSControl, _ suffixText: String? = nil) -> NSStackView {
+        let label = NSTextField(wrappingLabelWithString: labelText + ": ")
+        label.alignment = .right
+        label.widthAnchor.constraint(equalToConstant: calcLabelWidth()).isActive = true
+
+        control.identifier = NSUserInterfaceItemIdentifier(rawName)
```

Comment:
Aaah, I knew there was a way to embed some ID in these controls so that later they can be identified in the callback... I remember going through the different attributes without finding one that was a `String`. I see you store it in `NSUserInterfaceItemIdentifier` then later grab the `rawValue`. Very happy to see the horrible `inputsMap` go away 🙏

Now that you found how to ID the inputs, you should remove all attributes of this class! You can remove everything from line 10 to 19, as well as simplify all methods on lines 57 to 67. These members only existed because I needed to check which input changed in the callbacks but now it's useless. The parent view holds a reference so the object can be a local variable here that goes out of scope without having the object deallocated.

Reply:
Oh, you are absolutely right. Didn't come into my mind to doubt the existence of this view references. 👍